enhancements/ingress/route-subdomain.md

```diff
@@ -0,0 +1,642 @@
+  `spec.subdomain` is specified; this PR was never merged.
+* 2021-11-02,
+  [openshift/router#357](https://github.com/openshift/router/pull/357) was
+  opened, superseding #19;
+  [openshift/openshift-apiserver#254](https://github.com/openshift/openshift-apiserver/pull/254)
+  was opened, superseding
+  [openshift-apiserver#194](https://github.com/openshift/openshift-apiserver/pull/194);
+  and
+  [openshift/cluster-ingress-operator#674](https://github.com/openshift/cluster-ingress-operator/pull/674)
+  was opened to configure the domain in router deployments.
+* 2022-02-02, this enhancement was filed.
+
+## Drawbacks
+
+Implementing `spec.subdomain` expands the set of features that would need to be
+defined and implemented in any successor to the Route API and OpenShift router.
+In particular, [Gateway API's HTTPRoute
+kind](https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go)
+defines no equivalent API field; this gap could impede users from migrating from
+the Route API to Gateway API.
+
+## Alternatives
+
+### Require explicit `spec.host` and multiple Routes per domain (the status quo)
+
+For the use-case in which a user wants a custom subdomain and the cluster
+ingress domain, the user (human or software) can manually look up the cluster
+ingress domain or the default IngressController's domain and specify an explicit
+value for `spec.host` with the desired subdomain and the correct domain.
+Compared to using `spec.subdomain`, this alternative puts unnecessary
+operational burden on users.
+
+For the use-case in which the user wants multiple router deployments to expose a
+Route using each router's respective domain, the user would need to create
+multiple Routes: one per router deployment, explicitly specifying the
+appropriate domain for each Route in `spec.host`.  Again, this alternative puts
+unnecessary operational burden on users.
+
+### Define a mutating webhook to set `spec.host`
+
+For the sharding use-case, a custom mutating webhook or other controller could
+set `spec.host` using the appropriate domain based on the domain of the router
+deployment corresponding to the shard to which the Route belongs.  An advantage
+to this approach is that the controller could handle other aspects of sharding
+as well, such as automatically assign Routes to shards based on capacity or
+other factors (cf. [RFE-1467](https://issues.redhat.com/browse/RFE-1467) for an
+exploration of this idea).  Disadvantages of this approach include the
+additional development and operations effort required to implement and manage
+such a controller as well as the fundamental limitation that a Route could still
+only have a single host name and thus could not be simultaneously be exposed by
+multiple shards with different domains.
+
+### Default `spec.subdomain` instead of defaulting `spec.host`
+
+A possible modification to this enhancement would be to change the OpenShift API
```

Comment:
So my hot take is: I think this alternative is the most correct, as it paves a way to a single domain-specifier solution. Here's my arguments:
1. As a user, the relationship between `spec.host` and `spec.subdomain` (without reading the docs), is not natural to me like I mentioned in an comment above.
2. I just generally think the existing `spec.host` is a little redundant that you have to copy/paste the full domain into the route to match your ingress controller (hope you don't have copy/paste error). I'd think just `spec.subdomain` is a easier way to configure your domains.

So, I'm still new to enhancements and I'm not out-right disagreeing with this one, but I am curious if there is agreement that there is some discontinuity between the two, and if you really actually need both `spec.host` and `spec.subdomain`. 

The questions I have:
- In a perfect world, do we agree that we could really get by with just `spec.subdomain`?
- If we agree above, What's the transition plan to consolidate the two specifiers, if any?

Reply:
> So my hot take is: I think this alternative is the most correct, as it paves a way to a single domain-specifier solution. Here's my arguments:
> 
>     1. As a user, the relationship between `spec.host` and `spec.subdomain` (without reading the docs), is not natural to me like I mentioned in an comment above.

Agreed.  

>     2. I just generally think the existing `spec.host` is a little redundant that you have to copy/paste the full domain into the route to match your ingress controller (hope you don't have copy/paste error). I'd think just `spec.subdomain` is a easier way to configure your domains.

Disagree.  See below.  

> So, I'm still new to enhancements and I'm not out-right disagreeing with this one, but I am curious if there is agreement that there is some discontinuity between the two, and if you really actually need both `spec.host` and `spec.subdomain`.
> 
>     * In a perfect world, do we agree that we could really get by with just `spec.subdomain`?

No, the user may need to specify a host name outside of the ingress controller's domain.  It would be up to the user to create a DNS record pointing that custom host name to the ingress controller (typically, the user would use a CNAME record pointing to the `routerCanonicalHostname` from the route's status), but notwithstanding that, use cases do exist for specifying `spec.host` instead of `spec.subdomain`.  

The idea with this alternative is really that the user would still be able to specify `spec.host` or `spec.subdomain`, but if the user specified _neither_, the API would default `spec.subdomain`, rather than defaulting `spec.host`.  The next paragraph describes the tradeoff, which is, in short, that defaulting `spec.subdomain` instead of defaulting `spec.host` would be surprising and could break users' long-established workflows.

Reply:
Okay I see your point, I was unaware of specific use cases with `spec.host`, but that makes sense now. I generally agree with this position.

I still think the domain-related API is getting a little confusing, which I think you agree with. I'm curious, if we had a chance to redo it, what would be a more natural solution. That can be a future discussion.